third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -170,105 +233,219 @@ impl BorrowEdgeKind {
+    /// Joins two maps with labels in their range. For overlapping keys pointing to different labels,
+    /// the nodes behind the labels in the graph are joined, and the label in the other graph is
+    /// replaced by the given one.
+    ///
+    /// This function extends an active label renamings resulting from incrementally joining nodes.
     fn join_label_map<A: Clone + Ord>(
         &mut self,
         map: &mut BTreeMap<A, LifetimeLabel>,
         other_map: &BTreeMap<A, LifetimeLabel>,
-        mk_location: impl Fn(&A) -> MemoryLocation,
+        renaming: &mut BTreeMap<LifetimeLabel, LifetimeLabel>,
     ) -> JoinResult {
         let mut change = JoinResult::Unchanged;
         for (k, other_label) in other_map {
             match map.entry(k.clone()) {
                 btree_map::Entry::Vacant(entry) => {
-                    entry.insert(other_label.clone());
+                    entry.insert(*other_label);
                     change = JoinResult::Changed;
                 },
-                btree_map::Entry::Occupied(mut entry) => {
-                    let label = entry.get().clone();
-                    if &label != other_label && self.node(&label) != self.node(other_label) {
-                        // Create a new intermediate node and make it child of the other ones.
-                        let new_label = LifetimeLabel::derive(label.clone(), other_label.clone());
-                        self.new_node(new_label.clone(), mk_location(entry.key()));
-                        entry.insert(new_label.clone());
-                        // Determine mutability from the other nodes and add a `Skip` edge
-                        let is_mut = self
-                            .node(&label)
-                            .children
-                            .iter()
-                            .chain(self.node(other_label).children.iter())
-                            .any(|e| e.is_mut);
-                        let skip_edge =
-                            BorrowEdge::new(BorrowEdgeKind::Skip, is_mut, None, new_label.clone());
-                        self.add_edge(&label, skip_edge.clone());
-                        self.add_edge(other_label, skip_edge);
+                btree_map::Entry::Occupied(entry) => {
+                    let label = entry.get();
+                    if entry.get() != other_label {
+                        // Merge other node into this one, and add renaming of label.
+                        let other_copy = self.node(other_label).clone(); // can't mut and read same time
+                        self.node_mut(label).join(&other_copy);
+                        renaming.insert(*other_label, *label);
                         change = JoinResult::Changed;
                     }
                 },
             }
         }
         change
     }
+
+    fn rename_label(renaming: &BTreeMap<LifetimeLabel, LifetimeLabel>, label: &mut LifetimeLabel) {
+        while let Some(actual) = renaming.get(label) {
```

Comment:
I can't think of an example where repeating the replacement is needed here.  Any idea why this might be needed?

Reply:
As said, given that expectation is that graphs are small, I prefer for now the more readable solution, though I like your idea here (see it also on another PR)